charms/worker/k8s/src/upgrade.py

```diff
@@ -68,25 +63,52 @@ def __init__(self, charm: K8sCharmProtocol, cluster_inspector: ClusterInspector,
-        """Set the Juju upgrade status.
+    @property
+    def upgrade_granted(self) -> bool:
+        """Check if the upgrade has been granted."""
+        return self._upgrade_granted
+
+    def handler(self, event: ops.EventBase):
+        """Block reconciler if the unit was upgraded without a pre-upgrade-check.
 
         Args:
-            event: The UpdateStatusEvent instance.
+            event: ops.EventBase - event that triggered the check
         """
-        upgrade_status = self.state
-        if not upgrade_status:
-            return
-        if upgrade_status == "upgrading":
-            if not self.charm.is_upgrade_granted:
-                self._upgrade(event)
-        elif upgrade_status == "recovery":
-            status.add(ops.MaintenanceStatus("Charm is in recovery mode. Please check the logs."))
-            return
-        elif upgrade_status == "failed":
-            status.add(ops.BlockedStatus("Upgrade Failed. Please check the logs."))
-            return
+        if isinstance(event, (UpgradeGrantedEvent, UpgradeFinishedEvent)):
+            log.debug("%s, proceed with reconciliation", event.__class__.__name__)
+        elif self.state is None:
+            log.debug("Upgrade not setup, proceed with reconciliation")
+        elif self.state == "upgrading" and (failure := self._upgrade(event)):
+            log.warning("Upgrade failure, cease reconciliation: %s", failure.message)
+            status.add(failure)
+            raise status.ReconcilerError(failure.message)
+        elif self.state != "idle":
+            failure = self.charm.unit.status.message
+            message = f"Upgrade not idle state='{self.state}', cease reconciliation.: {failure}"
+            log.warning(message)
+            status.add(self.charm.unit.status)
+            raise status.ReconcilerError(message)
+        elif isinstance(event, ops.UpgradeCharmEvent):
+            # A resource attachment can trigger this event
+            if self.charm.snap_installation_resource.is_updated:
+                log.debug("Resource Attachment, proceed with reconciliation.")
+            # or it's a charm upgrade, maybe a single node upgrade that's done?
+            elif len(self.app_units) == 1 and self._upgrade_complete:
+                log.debug("Single node upgrade completed, proceed with reconciliation.")
+            # or maybe a multi-unit upgrade that is underway?
+            elif super().upgrade_stack:
+                log.debug("Upgrade stack exists, proceed with reconciliation.")
+            else:
+                reason = "Unit was upgraded without a pre-upgrade-check"
+                log.warning(reason)
+                self.set_unit_failed(reason)
+                if self.charm.unit.is_leader():
+                    self.upgrade_stack = self.build_upgrade_stack()
+                status.add(self.charm.unit.status)
+                raise status.ReconcilerError(reason)
```

Comment:
The bulk of the logic really went into this one function and is documented in the PR description